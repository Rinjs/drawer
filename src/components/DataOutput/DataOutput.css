.dataOutput {
  position: relative;
  min-height: 500px;
  width: 320px;
  text-align: center;
}

.dataOutput__button {
  position: absolute;
  right: 5px;
  top: 5px;
}

@media screen and (max-width: 800px) {
  .dataOutput {
    width: 290px;
  }
}

@media screen and (max-width: 600px) {
  .dataOutput {
    width: 250px;
  }
}

@media screen and (max-width: 400px) {
  .dataOutput {
    width: 220px;
  }
}
@media screen and (max-width: 320px) {
  .dataOutput {
    width: 200px;
  }
}
