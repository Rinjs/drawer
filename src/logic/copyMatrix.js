export const copyMatrix = matrix => [...matrix.map(item => [...item])];
