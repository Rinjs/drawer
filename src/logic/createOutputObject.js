export const createOutputObject = (
  text = "No data",
  errors = [],
  href = null
) => {
  return {
    text: text,
    errors: errors.length ? errors : ["No errors"],
    hasError: !!errors.length,
    href:
      href &&
      `data:text/plain;content-disposition=attachment;filename=file,${text}`
  };
};
