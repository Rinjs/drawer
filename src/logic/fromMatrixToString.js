export const fromMatrixToString = matrixArray => {
  let str = "";

  matrixArray.forEach(item => {
    item.forEach(element => (str += element.join("") + "\n"));
  });

  return str;
};
