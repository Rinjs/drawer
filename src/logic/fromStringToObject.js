export const fromStringToObject = (str = "") => {
  const commandsArr = str.split("\n");
  return str
    ? commandsArr.map(item => {
        return getCommands(item.trim().split(" "));
      })
    : {};
};

const getCommands = command => {
  const [commandType, ...commandValue] = command;

  return {
    [commandType]:
      commandType !== "B"
        ? [...commandValue.map(element => +element)]
        : [+commandValue[0], +commandValue[1], commandValue[2]]
  };
};
