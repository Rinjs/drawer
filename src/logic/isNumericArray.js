export const isNumericArray = array => {
  for (let i = 0; i < array.length; i++) {
    if (isNaN(parseFloat(array[i])) && !isFinite(array[i])) {
      return false;
    }
  }

  return true;
};
