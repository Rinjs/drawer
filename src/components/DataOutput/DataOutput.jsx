import React, { useState } from "react";

import "./DataOutput.css";
import { ErrorsList } from "../ErrorsList/ErrorsList";

export const DataOutput = ({ data }) => {
  const [buttonActive, setHasErrors] = useState(false);
  const buttonStyle = data.hasError
    ? "dataOutput__button btn btn-danger"
    : "dataOutput__button btn btn-secondary";

  return (
    <div
      className="shadow pt-5 p-2 mb-3 mt-3 bg-light rounded dataOutput"
      style={{
        minHeight: "500px",
        minWidth: "300px",
        maxWidth: "300px",
        textAlign: "center"
      }}
    >
      <button
        className={buttonStyle}
        onClick={() => setHasErrors(!buttonActive)}
      >
        Errors
      </button>
      {buttonActive ? (
        <ErrorsList errors={data.errors} />
      ) : (
        <pre>{data.text}</pre>
      )}
    </div>
  );
};
