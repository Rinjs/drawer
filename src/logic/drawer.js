import { copyMatrix } from "./copyMatrix";
import { checkError, isDotVisited, isValidCommand } from "./validation";
import { fromMatrixToString } from "./fromMatrixToString";
import { createOutputObject } from "./createOutputObject";

export const drawing = commands => {
  let canvasMatrixArray = [];
  const canvasParams = commands[0].C;
  const errorsArray = [];

  if (!commands[0].hasOwnProperty("C") || !isValidCommand("C", commands[0].C)) {
    errorsArray.push("Can't draw without canvas");

    return createOutputObject("No data", errorsArray);
  }

  commands.forEach(obj => {
    for (let key in obj) {
      switch (key) {
        case "C":
          canvasMatrixArray.push(createCanvas(...obj.C));
          break;
        case "L":
          const canvasWithLine = createLine(
            canvasMatrixArray[canvasMatrixArray.length - 1],
            ...obj.L
          );

          if (!canvasWithLine) {
            errorsArray.push(
              "Can’t draw a line, please check command and coordinates"
            );
            break;
          }
          if (isValidCommand("L", obj.L, canvasParams)) {
            canvasMatrixArray = checkError(canvasMatrixArray, canvasWithLine);
          } else
            errorsArray.push(
              "Can’t draw a line, please check command and coordinates"
            );

          break;
        case "R":
          if (isValidCommand("R", obj.R, canvasParams)) {
            canvasMatrixArray = checkError(
              canvasMatrixArray,
              createRectangle(
                canvasMatrixArray[canvasMatrixArray.length - 1],
                ...obj.R
              )
            );
          } else
            errorsArray.push(
              "Can’t draw a rectangle, please check command and coordinates"
            );
          break;
        case "B":
          if (isValidCommand("B", obj.B, canvasParams)) {
            canvasMatrixArray = checkError(
              canvasMatrixArray,
              bucketFill(
                canvasMatrixArray[canvasMatrixArray.length - 1],
                canvasParams,
                ...obj.B
              )
            );
          } else
            errorsArray.push(
              "Can’t fill the canvas, please check command and coordinates"
            );
          break;
        default:
          break;
      }
    }
  });

  const outputCanvas = fromMatrixToString(canvasMatrixArray);

  return createOutputObject(outputCanvas, errorsArray,true);
};

const bucketFill = (canvas, canvasParams, x, y, c) => {
  const canvasWithFill = copyMatrix(canvas);
  const targetSymbol = canvas[y][x];
  const visitedArray = [];
  const dotsToCheck = [{ x, y }];

  while (dotsToCheck.length > 0) {
    const { x, y } = dotsToCheck.pop();
    const isDotAvailable = !(
      x < 0 ||
      y < 0 ||
      x > canvasParams[0] ||
      y > canvasParams[1] ||
      isDotVisited(x, y, visitedArray)
    );

    if (isDotAvailable && canvasWithFill[y][x] === targetSymbol) {
      canvasWithFill[y][x] = c;

      visitedArray.push({ x, y });
      dotsToCheck.push(
        { x, y: y + 1 },
        { x, y: y - 1 },
        { x: x - 1, y },
        { x: x + 1, y }
      );
    }
  }

  return canvasWithFill;
};

const createRectangle = (canvas, x1, y1, x2, y2) => {
  let canvasWithRectangle = [];

  canvasWithRectangle = createHorizontalLine(canvas, x1, x2, y1);
  canvasWithRectangle = createHorizontalLine(canvasWithRectangle, x1, x2, y2);
  canvasWithRectangle = createVerticalLine(canvasWithRectangle, y1, y2, x1);
  canvasWithRectangle = createVerticalLine(canvasWithRectangle, y1, y2, x2);

  return canvasWithRectangle;
};

const createLine = (canvas, x1, y1, x2, y2) => {
  let canvasWithLine = [];
  const isVertical = x1 === x2;
  const isHorizontal = y1 === y2;

  if (isVertical) {
    canvasWithLine = createVerticalLine(canvas, y1, y2, x1);
  } else if (isHorizontal) {
    canvasWithLine = createHorizontalLine(canvas, x1, x2, y1);
  }

  return isVertical || isHorizontal ? canvasWithLine : false;
};

const createVerticalLine = (canvas, y1, y2, x) => {
  const canvasWithVerticalLine = copyMatrix(canvas);
  const symbol = "x";
  for (let i = 1; i < canvasWithVerticalLine.length - 1; i++) {
    if (i >= y1 && i <= y2) {
      canvasWithVerticalLine[i][x] = symbol;
    }
  }

  return canvasWithVerticalLine;
};

const createHorizontalLine = (canvas, x1, x2, y) => {
  const canvasWithHorizontalLine = copyMatrix(canvas);
  const symbol = "x";
  for (let i = 1; i < canvasWithHorizontalLine[0].length - 1; i++) {
    if (i >= x1 && i <= x2) {
      canvasWithHorizontalLine[y][i] = symbol;
    }
  }

  return canvasWithHorizontalLine;
};

const createCanvas = (width, height) => {
  let canvas = Array(height + 2)
    .fill(" ")
    .map(() => Array(width + 2).fill(" "));

  canvas[0] = canvas[canvas.length - 1].fill("-");

  for (let i = 1; i < canvas.length - 1; i++) {
    canvas[i][0] = canvas[i][width + 1] = "|";
  }

  return canvas;
};
