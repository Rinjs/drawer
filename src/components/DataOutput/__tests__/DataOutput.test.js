import React from "react";
import renderer from "react-test-renderer";
import { DataOutput } from "../DataOutput";
import { createOutputObject } from "../../../logic/createOutputObject";

test("renders without crashing", () => {
  const component = renderer.create(
    <DataOutput
      data={createOutputObject("No data", ["some error, some error"])}
    />
  );

  expect(component).toMatchSnapshot();
});

test("renders without crashing", () => {
  const component = renderer.create(<DataOutput data={createOutputObject()} />);
  expect(component).toMatchSnapshot();
});

test("renders without crashing", () => {
  const component = renderer.create(
    <DataOutput data={createOutputObject("some data ---11-1-", [], true)} />
  );
  expect(component).toMatchSnapshot();
});
