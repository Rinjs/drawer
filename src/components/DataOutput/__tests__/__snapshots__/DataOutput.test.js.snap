// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  className="shadow pt-5 p-2 mb-3 mt-3 bg-light rounded dataOutput"
  style={
    Object {
      "maxWidth": "300px",
      "minHeight": "500px",
      "minWidth": "300px",
      "textAlign": "center",
    }
  }
>
  <button
    className="dataOutput__button btn btn-danger"
    onClick={[Function]}
  >
    Errors
  </button>
  <pre>
    No data
  </pre>
</div>
`;

exports[`renders without crashing 2`] = `
<div
  className="shadow pt-5 p-2 mb-3 mt-3 bg-light rounded dataOutput"
  style={
    Object {
      "maxWidth": "300px",
      "minHeight": "500px",
      "minWidth": "300px",
      "textAlign": "center",
    }
  }
>
  <button
    className="dataOutput__button btn btn-secondary"
    onClick={[Function]}
  >
    Errors
  </button>
  <pre>
    No data
  </pre>
</div>
`;

exports[`renders without crashing 3`] = `
<div
  className="shadow pt-5 p-2 mb-3 mt-3 bg-light rounded dataOutput"
  style={
    Object {
      "maxWidth": "300px",
      "minHeight": "500px",
      "minWidth": "300px",
      "textAlign": "center",
    }
  }
>
  <button
    className="dataOutput__button btn btn-secondary"
    onClick={[Function]}
  >
    Errors
  </button>
  <pre>
    some data ---11-1-
  </pre>
</div>
`;
